git-server/src/lib.rs

```diff
@@ -0,0 +1,221 @@
+#![allow(clippy::type_complexity)]
+mod error;
+
+use std::collections::HashMap;
+use std::io::{BufRead, Read};
+use std::path::PathBuf;
+use std::process::{Command, Stdio};
+use std::{io, net};
+
+use http::{HeaderMap, Method};
+use warp::hyper::StatusCode;
+use warp::reply;
+use warp::{self, path, Buf, Filter, Rejection, Reply};
+
+use error::Error;
+
+pub const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[derive(Debug, Clone)]
+pub struct Options {
+    pub root: PathBuf,
+    pub listen: net::SocketAddr,
+    pub tls_cert: Option<PathBuf>,
+    pub tls_key: Option<PathBuf>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Context {
+    root: PathBuf,
+}
+
+/// Run the Git Server.
+pub async fn run(options: Options) {
+    let ctx = Context { root: options.root };
+    let server = warp::filters::any::any()
+        .map(move || ctx.clone())
+        .and(warp::method())
+        .and(warp::filters::header::headers_cloned())
+        .and(warp::filters::body::aggregate())
+        .and(warp::filters::addr::remote())
+        .and(path::tail())
+        .and(
+            warp::filters::query::raw()
+                .or(warp::any().map(String::default))
+                .unify(),
+        )
+        .and_then(git_handler)
+        .recover(recover)
+        .with(warp::log("radicle_git_server"));
+    let server = warp::serve(server);
+
+    if let (Some(cert), Some(key)) = (options.tls_cert, options.tls_key) {
+        server
+            .tls()
+            .cert_path(cert)
+            .key_path(key)
+            .run(options.listen)
+            .await
+    } else {
+        server.run(options.listen).await
+    }
+}
+
+async fn git_handler(
+    ctx: Context,
+    method: Method,
+    headers: HeaderMap,
+    body: impl Buf,
+    remote: Option<net::SocketAddr>,
+    path: warp::filters::path::Tail,
+    query: String,
+) -> Result<impl Reply, Rejection> {
+    let remote = remote.expect("there is always a remote for HTTP connections");
```

Comment:
Why make it an `Option` then?

Reply:
It's `warp` that gives me an `Option` :<

Claimed reason being: the transport may not have a remote addr :shrug:

Reply:
I'm guessing it's because they support like, unix domain socket as a transport? idk..

Reply:
Proof enough that warp isn't an HTTP server?

:trollface: